File: app/src/main/java/com/dumuzeyn/mp3player/DeviceAudioClassifier.java
package com.dumuzeyn.mp3player;

import java.util.Locale;

/** Conservative policy for media that can be added without an explicit user selection. */
final class DeviceAudioClassifier {
    private static final long MIN_MUSIC_DURATION_MS = 20_000L;
    private static final String[] VOICE_PATH_MARKERS = {
            "/recordings/", "/voice recorder/", "/voicerecorder/", "/sound_recorder/",
            "/call recordings/", "/callrecordings/", "/call_rec/", "/voice notes/",
            "/voicenotes/", "/whatsapp voice notes/", "/диктофон/"
    };
    private static final String[] VOICE_TEXT_MARKERS = {
            "voice recording", "voice record", "voice note", "audio recording",
            "call recording", "recorded call", "диктофон", "голосовая запись",
            "голосовое сообщение", "запись звонка"
    };

    private DeviceAudioClassifier() {
    }

    static boolean shouldAutoImport(Candidate candidate) {
        if (candidate == null || !candidate.music || candidate.recording
                || candidate.podcast || candidate.audiobook || candidate.ringtone
                || candidate.alarm || candidate.notification
                || candidate.durationMs < MIN_MUSIC_DURATION_MS) {
            return false;
        }
        String path = normalizedPath(candidate.relativePath);
        for (String marker : VOICE_PATH_MARKERS) {
            if (path.contains(marker)) {
                return false;
            }
        }
        String metadata = normalize(candidate.displayName) + ' ' + normalize(candidate.title)
                + ' ' + normalize(candidate.artist) + ' ' + normalize(candidate.album);
        for (String marker : VOICE_TEXT_MARKERS) {
            if (metadata.contains(marker)) {
                return false;
            }
        }
        if (looksLikeGeneratedVoiceMessage(candidate, path)) {
            return false;
        }
        return true;
    }

    private static boolean looksLikeGeneratedVoiceMessage(Candidate candidate, String path) {
        String file = normalize(candidate.displayName);
        if (file.matches("ptt-\\d{8}-wa\\d+\\.(opus|ogg|m4a|mp3)")) {
            return true;
        }
        boolean metadataUnknown = isUnknown(candidate.artist) && isUnknown(candidate.album);
        boolean generatedWhatsAppMetadata = isUnknown(candidate.artist)
                && "whatsapp audio".equals(normalize(candidate.album));
        if ((metadataUnknown || generatedWhatsAppMetadata) && path.contains("/whatsapp audio/")
                && file.matches("aud-\\d{8}-wa\\d+\\.(opus|ogg|m4a|mp3)")) {
            return true;
        }
        return metadataUnknown && (file.startsWith("voice_")
                || file.startsWith("voice-") || file.startsWith("recording_")
                || file.startsWith("recording-") || file.startsWith("call_record"));
    }

    private static boolean isUnknown(String value) {
        String normalized = normalize(value);
        return normalized.isEmpty() || "<unknown>".equals(normalized)
                || "unknown artist".equals(normalized)
                || "unknown album".equals(normalized);
    }

    private static String normalizedPath(String value) {
        String normalized = normalize(value).replace('\\', '/');
        return '/' + normalized.replaceAll("/+", "/") + '/';
    }

    private static String normalize(String value) {
        return value == null ? "" : value.toLowerCase(Locale.ROOT).trim();
    }

    static final class Candidate {
        final boolean music;
        final boolean recording;
        final boolean podcast;
        final boolean audiobook;
        final boolean ringtone;
        final boolean alarm;
        final boolean notification;
        final long durationMs;
        final String relativePath;
        final String displayName;
        final String title;
        final String artist;
        final String album;

        Candidate(boolean music, boolean recording, boolean podcast, boolean audiobook,
                boolean ringtone, boolean alarm, boolean notification, long durationMs,
                String relativePath, String displayName, String title, String artist,
                String album) {
            this.music = music;
            this.recording = recording;
            this.podcast = podcast;
            this.audiobook = audiobook;
            this.ringtone = ringtone;
            this.alarm = alarm;
            this.notification = notification;
            this.durationMs = durationMs;
            this.relativePath = relativePath;
            this.displayName = displayName;
            this.title = title;
            this.artist = artist;
            this.album = album;
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/DeviceAudioClassifier.kt
package com.dumuzeyn.mp3player

import java.util.Locale

/** Conservative policy for media that can be added without an explicit user selection. */
object DeviceAudioClassifier {
    private const val MIN_MUSIC_DURATION_MS = 20_000L
    private val voicePathMarkers = arrayOf(
        "/recordings/", "/voice recorder/", "/voicerecorder/", "/sound_recorder/",
        "/call recordings/", "/callrecordings/", "/call_rec/", "/voice notes/",
        "/voicenotes/", "/whatsapp voice notes/", "/диктофон/",
    )
    private val voiceTextMarkers = arrayOf(
        "voice recording", "voice record", "voice note", "audio recording",
        "call recording", "recorded call", "диктофон", "голосовая запись",
        "голосовое сообщение", "запись звонка",
    )
    private val pttPattern = Regex("ptt-\\d{8}-wa\\d+\\.(opus|ogg|m4a|mp3)")
    private val audioPattern = Regex("aud-\\d{8}-wa\\d+\\.(opus|ogg|m4a|mp3)")

    @JvmStatic
    fun shouldAutoImport(candidate: Candidate?): Boolean {
        if (
            candidate == null || !candidate.music || candidate.recording || candidate.podcast ||
            candidate.audiobook || candidate.ringtone || candidate.alarm || candidate.notification ||
            candidate.durationMs < MIN_MUSIC_DURATION_MS
        ) return false
        val path = normalizedPath(candidate.relativePath)
        if (voicePathMarkers.any(path::contains)) return false
        val metadata = listOf(
            candidate.displayName,
            candidate.title,
            candidate.artist,
            candidate.album,
        ).joinToString(" ", transform = ::normalize)
        if (voiceTextMarkers.any(metadata::contains)) return false
        return !looksLikeGeneratedVoiceMessage(candidate, path)
    }

    private fun looksLikeGeneratedVoiceMessage(candidate: Candidate, path: String): Boolean {
        val file = normalize(candidate.displayName)
        if (pttPattern.matches(file)) return true
        val metadataUnknown = isUnknown(candidate.artist) && isUnknown(candidate.album)
        val generatedWhatsAppMetadata = isUnknown(candidate.artist) &&
            normalize(candidate.album) == "whatsapp audio"
        if (
            (metadataUnknown || generatedWhatsAppMetadata) && "/whatsapp audio/" in path &&
            audioPattern.matches(file)
        ) return true
        return metadataUnknown && listOf(
            "voice_", "voice-", "recording_", "recording-", "call_record",
        ).any(file::startsWith)
    }

    private fun isUnknown(value: String?): Boolean = normalize(value).let {
        it.isEmpty() || it == "<unknown>" || it == "unknown artist" || it == "unknown album"
    }

    private fun normalizedPath(value: String?): String =
        "/" + normalize(value).replace('\\', '/').replace(Regex("/+"), "/") + "/"

    private fun normalize(value: String?): String = value?.lowercase(Locale.ROOT)?.trim().orEmpty()

    class Candidate(
        @JvmField val music: Boolean,
        @JvmField val recording: Boolean,
        @JvmField val podcast: Boolean,
        @JvmField val audiobook: Boolean,
        @JvmField val ringtone: Boolean,
        @JvmField val alarm: Boolean,
        @JvmField val notification: Boolean,
        @JvmField val durationMs: Long,
        @JvmField val relativePath: String?,
        @JvmField val displayName: String?,
        @JvmField val title: String?,
        @JvmField val artist: String?,
        @JvmField val album: String?,
    )
}
